### PR TITLE
Add Fastapi server to serve TTS and download script

### PR DESCRIPTION
**Pull request by iflamed**
1. Add fastapi server in the `main.py`
2. Create `download.py` to download models
3. Update readme to show how to use fastapi.

### CODE REVIEW

**Comment by iflamed** [comment]
@aluminumbox You may review this PR, and then merge it.

**Comment by lonngxiang** [comment]
能把这些接口函数都支持吗  
image

**Comment by iflamed** [comment]
> 能把这些接口函数都支持吗 image

@lonngxiang 我正在改，一点点来

**Comment by lonngxiang** [comment]
我这边用flask实现了4个接口,测试了前3个正常运行，第四个模型没下载暂时没测试

api.py

```

import time
import io, os, sys
ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append('{}'.format(ROOT_DIR))
sys.path.append('{}/third_party/Matcha-TTS'.format(ROOT_DIR))

import numpy as np
from flask import Flask, request, Response
import torch
import torchaudio

from cosyvoice.cli.cosyvoice import CosyVoice
from cosyvoice.utils.file_utils import load_wav

##指定模型地址,对于零样本/跨语言推理，请使用CosyVoice-300M模型。对于SFT推理，
##请使用CosyVoice-300M-SFT模型。对于指令推理，请使用CosyVoice-300M-Instruct模型。
cosyvoice = CosyVoice('/data/pretrained_models/CosyVoice-300M') 

print(cosyvoice.list_avaliable_spks())

app = Flask(__name__)

@app.route("/inference/sft", methods=['POST'])
def sft():
    question_data = request.get_json()
    query = question_data.get('query')
    speaker = question_data.get('speaker')
    if not query:
        return {"error": "Query parameter 'query' is required"}, 400

    start = time.process_time()
    output = cosyvoice.inference_sft(query, speaker)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")


@app.route("/inference/zero_shot", methods=['POST'])
def zero_shot():
    question_data = request.get_json()

    tts_text = question_data.get('query')
    prompt_text = question_data.get('prompt_text')
    prompt_speech = load_wav(question_data.get('prompt_speech'), 16000)
    prompt_audio = (prompt_speech.numpy() * (2**15)).astype(np.int16).tobytes()

    prompt_speech_16k = torch.from_numpy(np.array(np.frombuffer(prompt_audio, dtype=np.int16))).unsqueeze(dim=0)
    prompt_speech_16k = prompt_speech_16k.float() / (2**15)

    start = time.process_time()
    output = cosyvoice.inference_zero_shot(tts_text, prompt_text, prompt_speech_16k)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")


@app.route("/inference/cross_lingual", methods=['POST'])
def cross_lingual():
    question_data = request.get_json()

    tts_text = question_data.get('query')
    prompt_speech = load_wav(question_data.get('prompt_speech'), 16000)
    prompt_audio = (prompt_speech.numpy() * (2**15)).astype(np.int16).tobytes()

    prompt_speech_16k = torch.from_numpy(np.array(np.frombuffer(prompt_audio, dtype=np.int16))).unsqueeze(dim=0)
    prompt_speech_16k = prompt_speech_16k.float() / (2**15)

    start = time.process_time()
    output = cosyvoice.inference_cross_lingual(tts_text, prompt_speech_16k)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")


@app.route("/inference/instruct", methods=['POST'])
def instruct():
    question_data = request.get_json()
    tts_text = question_data.get('query')
    speaker = question_data.get('speaker')
    instruct_text = question_data.get('instruct_text')

    start = time.process_time()
    output = cosyvoice.inference_instruct(tts_text, speaker, instruct_text)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")
    


    

if __name__ == "__main__":
    app.run(host='0.0.0.0', port=50000)

```

**Comment by iflamed** [comment]
好啊，我把你代码合并到pr里面吧，一起推上去Sent from newbing’s iPhoneOn Jul 10, 2024, at 14:40, dragon10 ***@***.***> wrote:﻿
我这边用flask实现了4个接口,测试了前3个正常运行，第四个模型没下载暂时没测试
api.py

import time
import io, os, sys
ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append('{}'.format(ROOT_DIR))
sys.path.append('{}/third_party/Matcha-TTS'.format(ROOT_DIR))

import numpy as np
from flask import Flask, request, Response
import torch
import torchaudio

from cosyvoice.cli.cosyvoice import CosyVoice
from cosyvoice.utils.file_utils import load_wav

##指定模型地址,对于零样本/跨语言推理，请使用CosyVoice-300M模型。对于SFT推理，
##请使用CosyVoice-300M-SFT模型。对于指令推理，请使用CosyVoice-300M-Instruct模型。
cosyvoice = CosyVoice('/data/pretrained_models/CosyVoice-300M') 

print(cosyvoice.list_avaliable_spks())

app = Flask(__name__)

@app.route("/inference/sft", methods=['POST'])
def sft():
    question_data = request.get_json()
    query = question_data.get('query')
    speaker = question_data.get('speaker')
    if not query:
        return {"error": "Query parameter 'query' is required"}, 400

    start = time.process_time()
    output = cosyvoice.inference_sft(query, speaker)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")


@app.route("/inference/zero_shot", methods=['POST'])
def zero_shot():
    question_data = request.get_json()

    tts_text = question_data.get('query')
    prompt_text = question_data.get('prompt_text')
    prompt_speech = load_wav(question_data.get('prompt_speech'), 16000)
    prompt_audio = (prompt_speech.numpy() * (2**15)).astype(np.int16).tobytes()

    prompt_speech_16k = torch.from_numpy(np.array(np.frombuffer(prompt_audio, dtype=np.int16))).unsqueeze(dim=0)
    prompt_speech_16k = prompt_speech_16k.float() / (2**15)

    start = time.process_time()
    output = cosyvoice.inference_zero_shot(tts_text, prompt_text, prompt_speech_16k)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")


@app.route("/inference/cross_lingual", methods=['POST'])
def cross_lingual():
    question_data = request.get_json()

    tts_text = question_data.get('query')
    prompt_speech = load_wav(question_data.get('prompt_speech'), 16000)
    prompt_audio = (prompt_speech.numpy() * (2**15)).astype(np.int16).tobytes()

    prompt_speech_16k = torch.from_numpy(np.array(np.frombuffer(prompt_audio, dtype=np.int16))).unsqueeze(dim=0)
    prompt_speech_16k = prompt_speech_16k.float() / (2**15)

    start = time.process_time()
    output = cosyvoice.inference_cross_lingual(tts_text, prompt_speech_16k)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")


@app.route("/inference/instruct", methods=['POST'])
def instruct():
    question_data = request.get_json()
    tts_text = question_data.get('query')
    speaker = question_data.get('speaker')
    instruct_text = question_data.get('instruct_text')

    start = time.process_time()
    output = cosyvoice.inference_instruct(tts_text, speaker, instruct_text)
    end = time.process_time()
    print("infer time:", end - start)
    buffer = io.BytesIO()
    torchaudio.save(buffer, output['tts_speech'], 22050, format="wav")
    buffer.seek(0)
    return Response(buffer.read(), mimetype="audio/wav")
    


    

if __name__ == "__main__":
    app.run(host='0.0.0.0', port=50000)

**Comment by aluminumbox** [comment]
is this PR ready?I believe we still need a fastapi_client.py, please add 铝箱 in Dingding group to discuss details

**Comment by iflamed** [comment]
> is this PR ready?I believe we still need a fastapi_client.py, please add 铝箱 in Dingding group to discuss details

Not ready!

**Comment by iflamed** [comment]
@aluminumbox Now, it's ready to merge.
